### PR TITLE
Docs now built with sphinx directly and API reference refactored

### PR DESCRIPTION
**Pull request by maximlt**
Main changes:

- `nbsite` is no longer the executable that builds the docs, instead it's done with sphinx directly. To make this change, I've had to move all the notebooks from `./examples` to `./doc`. This isn't a problem for the Param project as it doesn't have an `examples` command like Panel and other HoloViz projects. This change should make it easier to work on the docs and use tools like `sphinx-autobuild`.
- attempts to improve the API reference. I've played around with sphinx's `autosummary` extension, I'm not sure I like it entirely as you can't configure that much the autogenerated doc for each entry compared to what you can do with e.g. `autoclass`. It still feels like an improvement though!

Here are two ways to reference the API from the docs (thanks to this [SO post](https://stackoverflow.com/questions/74576125/refering-to-a-sphynx-autodoc-generated-python-method-pymeth-from-a-myst-p))

https://github.com/holoviz/param/blob/5d0cffbddadb85e04850e26f2bb1258f15aa501c/doc/reference/param.md?plain=1#L22
https://github.com/holoviz/param/blob/5d0cffbddadb85e04850e26f2bb1258f15aa501c/doc/user_guide/Parameter_Types.ipynb#L65

EDIT: I built the dev docs from this branch, see the new API page https://holoviz-dev.github.io/param/reference/index.html

@droumis do you mind reviewing this PR?

### CODE REVIEW

File: doc/reference/param.md
Comment (maximlt) [reply doc/reference/param.md:100]: Yes true. Though because the Parameters have pretty bare docstrings and Jim wrote https://param.holoviz.org/user_guide/Parameter_Types.html as a reference guide it isn't going to be clear what the reference source is going to be for the Parameters even with an update API page.

File: doc/reference/index.md
Comment (maximlt) [reply doc/reference/index.md:10]: True!

File: doc/reference/param.md
Comment (droumis) [reply doc/reference/param.md:100]: I think we should just stick to the idea that technical reference material should be found in API reference... and someday maybe the user guide pages will become more how-to-like so there's a more clear distinction

File: doc/reference/index.md
Comment (philippjfr) [inline doc/reference/index.md:10]:
```suggestion
```

**Review by philippjfr — APPROVED** [review]
I think this PR is a big improvement on what we had before so I'm going ahead and merging it. We also discussed this in our docs sync the other day and decided that manually curating overview pages is going to be the only sensible way forward but that we would likely try to add some automation to automate the process of listing all objects of a single type, e.g. for listing all Parameter subclasses, and for listing all methods/attributes of a specific class, e.g. for the Parameters namespace.

**Comment by droumis** [comment]
ok, I'm experimenting with ways to avoid the redirect and have a better tiered left nav, but I can finish later in another PR

**Comment by droumis** [comment]
Or, if you can wait until tomorrow to merge, I have some time later today

**Comment by philippjfr** [comment]
Sounds great, I'll wait.

**Comment by droumis** [comment]
I figured out how we can have the reference API section headings in the leftnav instead of the individual class pages. https://holoviz-dev.github.io/param/reference/index.html


https://github.com/holoviz/param/assets/6613202/374ca8c3-404a-4463-824a-c78f8a4c1243


In order to have links in the leftnav, each section heading has to be a separate file, so I've broken them out into files that are now in doc/reference/param/. However, to get their contents aggregated on the same page as before, I used the ['include' directive ](https://myst-parser.readthedocs.io/en/latest/syntax/organising_content.html#id4) to insert the content from each of the new section pages into a single param/index.md page. Additionally, to be listed in the leftnav, each page needs a title (e.g. '# title').. which screws up the include directive. To fix that, I used ":start-line: 2" for each include directive to skip over the title in the first line. 

Additionally, the autosummary extension automatically adds its generated pages to the toc... so in order to suppress this, I installed a new plugin called [sphinx-remove-toctrees](https://github.com/executablebooks/sphinx-remove-toctrees). This should also speed up the docs build for bigger holoviz projects if we choose to implement it across packages. I've added it to the 'doc' list in pyproject.toml for now.


Lastly, I've renamed "Main objects" to "Parameterized objects", and I've moved the `ParamOverrides` class from this section into the "Parameter helpers" section. I was trying to decide whether to place it in "Parameter helpers" or "Parameterized helpers", but considering the functionality of ParamOverrides, it seems more related to the manipulation and management of parameters, rather than event handling or watchers.

**Comment by droumis** [comment]
ready to merge. We can iterate on the structure more later

**Pull request merged**
